subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -379,13 +382,31 @@ private Configuration createApiElements(SourceSet mainSourceSet, PublishArtifact
                 .providesApi()
                 .withDescription("API elements for main."));
         apiElementsConfiguration.deprecateForDeclaration(IMPLEMENTATION_CONFIGURATION_NAME, COMPILE_ONLY_CONFIGURATION_NAME);
+        apiElementsConfiguration.getAttributes().attribute(CompileView.VIEW_ATTRIBUTE, objectFactory.named(CompileView.class, CompileView.JAVA_API));
 
         // Configure variants
         addJarArtifactToConfiguration(apiElementsConfiguration, jarArtifact);
 
         return apiElementsConfiguration;
     }
 
+    private void createCompileElements(Project project, SourceSet mainSourceSet, PublishArtifact jarArtifact) {
+        ConfigurationContainer configurations = project.getConfigurations();
+        Configuration implementationConfiguration = configurations.getByName(IMPLEMENTATION_CONFIGURATION_NAME);
+        Configuration compileOnly = configurations.getByName(COMPILE_ONLY_CONFIGURATION_NAME);
+
+        final DeprecatableConfiguration compileElementsConfiguration = (DeprecatableConfiguration) jvmServices.createOutgoingElements(COMPILE_ELEMENTS_CONFIGURATION_NAME,
```

Comment:
Good points here. I was just following the conventions for creating the other configurations in this file, but it makes sense that since this is a new configuration we can make it consumable and non-resolvable by default. 

I'm curious though, this method is called `deprecateForDeclaration`, though there is no way to call something like `.canBeDeclaredAgainst(false)`, only deprecation. Does it make sense to deprecate for declaration AND make it consumable and non-resolvable?

Reply:
I think it definitely makes sense to set the consumable and resolvable flags correctly, regardless of any deprecation concerns.